cpp/src/io/parquet/delta_binary.cuh

```diff
@@ -0,0 +1,281 @@
+    cur_mb_start = nullptr;
+
+    if (current_value_idx < num_encoded_values(is_decode)) {
+      auto d_start  = block_start;
+      cur_min_delta = get_zz128(d_start, block_end);
+      cur_bitwidths = d_start;
+
+      d_start += mini_block_count;
+      cur_mb_start = d_start;
+    }
+  }
+
+  // read delta binary header into state object. should be called on thread 0. header format is:
+  //
+  // | block size (uint) | mini-block count (uint) | value count (uint) | first value (int) |
+  //
+  // also initializes the first mini-block before exit
+  inline __device__ void init_binary_block(uint8_t const* d_start, uint8_t const* d_end)
+  {
+    block_end        = d_end;
+    block_size       = get_uleb128(d_start, d_end);
+    mini_block_count = get_uleb128(d_start, d_end);
+    value_count      = get_uleb128(d_start, d_end);
+    last_value       = get_zz128(d_start, d_end);
+
+    current_value_idx = 0;
+    values_per_mb     = block_size / mini_block_count;
+
+    // init the first mini-block
+    block_start = d_start;
+    init_mini_block(false);
+  }
+
+  // skip to the start of the next mini-block. should only be called on thread 0.
+  // calls init_binary_block if currently on the last mini-block in a block.
+  inline __device__ void setup_next_mini_block(bool is_decode)
+  {
+    if (current_value_idx >= num_encoded_values(is_decode)) { return; }
+
+    current_value_idx += values_per_mb;
+
+    // just set pointer to start of next mini_block
+    if (cur_mb < mini_block_count - 1) {
+      cur_mb_start += cur_bitwidths[cur_mb] * values_per_mb / 8;
+      cur_mb++;
+    }
+    // out of mini-blocks, start a new block
+    else {
+      block_start = cur_mb_start + cur_bitwidths[cur_mb] * values_per_mb / 8;
+      init_mini_block(is_decode);
+    }
+  }
+
+  // decode the current mini-batch of deltas, and convert to values.
+  // called by all threads in a warp, currently only one warp supported.
+  inline __device__ void calc_mini_block_values(int lane_id)
+  {
+    using cudf::detail::warp_size;
+    if (current_value_idx >= value_count) { return; }
+
+    // need to save first value from header on first pass
+    if (current_value_idx == 0) {
+      if (lane_id == 0) {
+        current_value_idx++;
+        value[0] = last_value;
+      }
+      __syncwarp();
+    }
+
+    uint32_t const mb_bits = cur_bitwidths[cur_mb];
+
+    // need to do in multiple passes if values_per_mb != 32
+    uint32_t const num_pass = values_per_mb / warp_size;
+
+    auto d_start = cur_mb_start;
+
+    for (int i = 0; i < num_pass; i++) {
+      // position at end of the current mini-block since the following calculates
+      // negative indexes
+      d_start += (warp_size * mb_bits) / 8;
+
+      // unpack deltas. modified from version in gpuDecodeDictionaryIndices(), but
+      // that one only unpacks up to bitwidths of 24. simplified some since this
+      // will always do batches of 32. also replaced branching with a loop.
```

Comment:
I'll have to dig up old commits.  I used to have the loop unrolled...but didn't notice a huge difference at the time.  I'll revisit.

Reply:
I tried replacing the loop with my old unrolled version:
```c++
          if (c < mb_bits && p < block_end) {  // up to 8 bits
            delta |= static_cast<zigzag128_t>(*p++) << c;
            c += 8;
            if (c < mb_bits && p < block_end) {  // up to 16 bits
              delta |= static_cast<zigzag128_t>(*p++) << c;
              c += 8;
              if (c < mb_bits && p < block_end) {  // 24 bits
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                if (c < mb_bits && p < block_end) {  // 32 bits
                  delta |= static_cast<zigzag128_t>(*p++) << c;
                  c += 8;
                  if (c < mb_bits && p < block_end) {  // 40 bits
                    delta |= static_cast<zigzag128_t>(*p++) << c;
                    c += 8;
                    if (c < mb_bits && p < block_end) {  // 48 bits
                      delta |= static_cast<zigzag128_t>(*p++) << c;
                      c += 8;
                      if (c < mb_bits && p < block_end) {  // 56 bits
                        delta |= static_cast<zigzag128_t>(*p++) << c;
                        c += 8;
                        if (c < mb_bits && p < block_end) {  // 64 bits
                          delta |= static_cast<zigzag128_t>(*p++) << c;
                        }
                      }
                    }
                  }
                }
              }
            }
          }
```
I also tried a version that uses a switch rather than nested branching:
```c++
          auto const bytes_left = (mb_bits - ofs + 7) / 8;
          if (p + bytes_left <= block_end) {
            switch (bytes_left) {
              case 8:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 7:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 6:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 5:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 4:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 3:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 2:
                delta |= static_cast<zigzag128_t>(*p++) << c;
                c += 8;
                [[fallthrough]];
              case 1: delta |= static_cast<zigzag128_t>(*p++) << c; [[fallthrough]];
              default: break;
            }
          }
```
I ran each of those through a 10M line file with columns of random int32 and int64. The int64 deltas should get pretty large and exercise the deeper parts of the unpacker.  In the end, the `while` and unrolled `if`s were pretty much a wash, as I remembered.  The switch version was a bit slower :(

Reply:
omg, that code. Good to see that the loop is no worse than other options.
Thanks for checking this.

Reply:
do you think the unrolled is worth the register savings (48 vs 55)?

Reply:
I don't think it is. Feels like a "gaze into the abyss" sort of ordeal :D
Could be persuaded otherwise with a benchmark, though :shrug:

Reply:
ok, I'll leave it alone then.  will just add a note.